### ISSUE TITLE
No tests found if `__tests__` directory is in the root

### ISSUE THREAD

**Issue by h0adp0re** [marilari88/neotest-vitest#22]
We have a `__tests__` folder in the root of our project for legacy reasons and the actual tests for our components are in the components' sub-folders within `src`. How to proceed?

**Comment by marilari88** [comment]
Hi @h0adp0re! I tried to repro your case and everything works fine (as you can see in the picture below). Can you tell me if you are experiencing any problems discovering the tests or running them? Do you see the tests in the summary panel? The order in which the adapter checks for the existence of the __tests__ folder should not affect the result of the is_test_file function.

![image](https://user-images.githubusercontent.com/32909388/224832183-1e969882-ee66-4f17-93cd-1da7b8c62730.png)

**Comment by h0adp0re** [comment]
Hey, that's strange. I have problems running them, but the summary panel actually does show 2 directories — `__tests__` and `src`. I tested a bit and here are some observations:
- Running a test in a test file via `require("neotest").run.run()` doesn't seem to do anything, although the editor slows down a bit for a few seconds, something must be happening in the background.
- None of the directories/suites/tests are expanded in the summary panel when opened while in a test file.
- When I expand the directories manually, none of the tests in a file are identified/listed in the summary.
- Only the test file is listed and running the file (with `summary.mappings.run`) results in the file passing successfully.
- The icon for the successfully run test file is "skipped".

After the initial run via the summary panel comes the interesting part:
- Saving the test file buffer will cause neotest symbols to render in the gutter for every test, which indicates that something just started working.
- Running tests via `require("neotest").run.run()` works as expected.
- Moving back to the summary panel, every single test is shown under the test file.
- Re-running the file will render the "passed" icon for the file and every suite and test under it.

Everything also works as expected when I delete the `__tests__` directory from root. 🤷‍♂️

Looks like some initialization problem? Because either deleting `__tests__` from root or running any test file via the summary panel first is the only way my neotest installation starts working at the moment.

**Comment by h0adp0re** [comment]
Alright well, thanks @AlexandrosAlexiou! #43 helped fix this issue for me. I simply told `neotest-vitest` to only consider what it finds in the `src` folder as tests in my special project.

Although I would like to confirm that I'm using this correctly. Is the `true` default sensible?

```lua
adapters = {
  require("neotest-vitest")({
    is_test_file = function(file_path)
      if string.match(file_path, "my-project") then
        return string.match(file_path, "/src/")
      end

      return true
    end,
  }),
},
```

And more, reading the source suggests that this check runs in addition to the usual checks?
Would moving this condition as the first one in the routine save some needless parsing in unrelated directories?
https://github.com/marilari88/neotest-vitest/blob/c0ea475596483eb02fa8e92c6be65c0536d55630/lua/neotest-vitest/init.lua#L430-L435

**Comment by AbaoFromCUG** [comment]
I found a practical way to make `is_test_file` more accurate when test files are not easy to detect by filename patterns.

Instead of relying on glob patterns, we can ask Vitest directly for the test file list:

```lua
require("neotest-vitest")({
  ---@async
  is_test_file = function(file_path)
    local nio = require("nio")

    local process = nio.process.run({
      cmd = "bun",
      args = { "vitest", "list", "--json" },
    })

    process.result()

    local test_cases = vim.json.decode(process.stdout.read())
    local test_files = vim.iter(test_cases)
      :map(function(test)
        return test.file
      end)
      :unique()
      :totable()

    return vim.tbl_contains(test_files, file_path)
  end,
})
```

This is useful when test files are determined by Vitest configuration rather than naming conventions.

One caveat: this runs bun vitest list --json every time is_test_file is called, so it may be worth caching the result if performance becomes an issue.


So, my final solution is 

```lua

---@generic T
---@param fn fun(...): T
---@param opts? { ttl_ms?: integer }
---@return async fun(...): T
local function singleflight(fn, opts)
    local nio = require("nio")
    opts = opts or {}

    local ttl_ms = opts.ttl_ms

    local future = nil
    local cached = nil
    local expires = 0

    local function now_ms()
        return vim.uv.hrtime() / 1e6
    end

    ---@async
    return function(...)
        local args = { ... }

        -- cache hit
        if ttl_ms and cached and now_ms() < expires then
            return cached
        end

        -- request coalescing
        if future then
            return future:wait()
        end

        local f = nio.control.future()
        future = f

        nio.run(function()
            return fn(unpack(args))
        end, function(ok, result)
            if ok and ttl_ms then
                cached = result
                expires = now_ms() + ttl_ms
            end


            if ok then
                f.set(result)
            else
                f.set_error(result)
            end
            if future == f then
                future = nil
            end
        end)
        return f.wait()
    end
end

local function setup_neotest()
    ---@async
    local function load_test_files()
        local nio = require("nio")
        local process = assert(nio.process.run({
            cmd = "bun",
            args = { "vitest", "list", "--json" },
        }))
        process.result()
        return vim.iter(vim.json.decode(process.stdout.read())):map(function(test)
            return test.file
        end):unique():totable()
    end
    -- cache with expire time 10s
    local get_test_files = singleflight(
        load_test_files,
        { ttl_ms = 10000 }
    )

    ---@diagnostic disable-next-line: missing-fields
    require("neotest").setup({
        adapters = {
           
            require("neotest-vitest")({
                ---@async
                is_test_file = function(file_path)
                    return vim.list_contains(get_test_files(), file_path)
                end,
            }),
         
    })
end
```


Could I open a PR to support dynamic test files/cases discovery via vitest list? @marilari88

**Comment by marilari88** [comment]
Hey @AbaoFromCUG, that would be great! 
How would you change the library's API contract?

**Comment by AbaoFromCUG** [comment]
### **`neotest-playwright`'s solution**
As a reference, [neotest-playwright](https://github.com/thenbe/neotest-playwright) already provides a configuration option:

```lua
require("neotest-playwright").adapter({
    options = {
        enable_dynamic_test_discovery = false,

        -- Custom criteria for a file path to be a test file. Useful in large
        -- projects or projects with peculiar tests folder structure. IMPORTANT:
        -- When setting this option, make sure to be as strict as possible. For
        -- example, the pattern should not return true for jpg files that may end up
        -- in your test directory.
        is_test_file = function(file_path)
            return ...
        end,
    },
})
```

As mentioned in the documentation for dynamic test discovery:

> **Dynamic Test Discovery**
> `neotest-playwright` can make use of the playwright cli to unlock extra features. Most importantly, the playwright cli provides information about which tests belongs to which project. neotest-playwright will parse this information to display, run, and report the results of tests on a per-project basis.
> 
> To enable this, set `enable_dynamic_test_discovery` to true.
> 
> Caveats
> This feature works by calling `playwright test --list --reporter=json`. While this is a relatively fast operation, it does add some overhead. Therefore, neotest-playwright only calls this feature once (when the adapter is first initialized). From then on, neotest-playwright continues to rely on treesitter to track your tests and enhance them with the data previously resolved by the playwright cli. There are times, however, where we want to refresh this data. To remedy this: neotest-playwright exposes a command `:NeotestPlaywrightRefresh`. This comes in handy in the following scenarios:
> 
> * Adding a new test
> * Renaming a test
> * Changing the project(s) configuration in your playwright.config.ts file

Their approach is to perform dynamic discovery only once during initialization, and expose a manual refresh command afterwards.

### My proposal

However, I think using something like `singleflight` to deduplicate concurrent refresh/discovery requests might provide a better UX overall. My proposal would be:
* The default implement of `is_test_file` is `vitest list  --json` with singleflight
* Expose a `discovery_ttl_ms` as the tts of `vitest list --json`

The options whould be 
```lua
---@class Option
---@field is_test_file async fun(file_path: string): boolean # checks if a file is a test file via `vitest list`, cached with 3s TTL using `singleflight` for concurrent request coalescing
---@field discovery_ttl_ms integer # cache TTL in milliseconds for discovery operations (default: 3s)

```

### Discussion

Compared to a one-time discovery + manual refresh command:

* users do not need to remember an extra command
* discovery can still stay lazy/on-demand
* repeated scans can be naturally throttled/deduplicated
* it feels more ergonomic, especially in large repos where test structure changes frequently

In other words, this shifts the complexity from the user to the implementation, which is usually a worthwhile tradeoff for developer tooling.


But ultimately, it’s entirely up to you. ❤️  @marilari88

**Comment by h0adp0re** [comment]
I tried `npx vitest list --json` in a larger project and it took 22 seconds. Is that ok?

**Comment by AbaoFromCUG** [comment]
> I tried `npx vitest list --json` in a larger project and it took 22 seconds. Is that ok?

Could you try running it with  local Vitest binary from `node_modules/.bin` instead of `npx`?

`npx` may add some overhead due to package resolution/installation, so this would help determine whether the 22-second runtime is coming from Vitest itself or from `npx`.

Also, note that users can provide a custom `is_test_file` implementation to override the default test file detection logic.

**Comment by AbaoFromCUG** [comment]
By the way, a better performance option could be `vitest  list --filesOnly --json`, it would avoid load/parse test files


For example, I tested this on the [Analog repository](https://github.com/analogjs/analog?utm_source=chatgpt.com), which contains 296 test files, and vitest list --filesOnly --json completed in about 0.3 seconds.

```
~/Documents/github/analog$ time bash  ./node_modules/.bin/vitest  list --filesOnly --json |wc -l
296
bash ./node_modules/.bin/vitest list --filesOnly --json  0.29s user 0.05s system 123% cpu 0.275 total
wc -l  0.00s user 0.00s system 1% cpu 0.274 total
```
